src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,551 @@
+            {
+                return null;
+            }
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEcsResponse(string relativeUri)
+        {
+            if (relativeUri == null)
+            {
+                return null;
+            }
+
+            var response = AwsHttpClientHelper.GetECSResponse(relativeUri).GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+            ValidateCredentials(username, password, sessionToken);
+
+            if (username == null)
+            {
+                return null;
+            }
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2Response().GetAwaiter().GetResult();
+            var parsedReponse = BsonDocument.Parse(response);
+            var username = parsedReponse.GetValue("AccessKeyId")?.AsString;
+            var password = parsedReponse.GetValue("SecretAccessKey")?.AsString;
+            var sessionToken = parsedReponse.GetValue("Token")?.AsString;
+            ValidateCredentials(username, password, sessionToken);
+
+            if (username == null)
+            {
+                return null;
+            }
+
+            return new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            string sessionToken = null;
+            if (properties != null)
+            {
+                foreach (var pair in properties)
+                {
+                    switch (pair.Key.ToUpperInvariant())
+                    {
+                        case "AWS_SESSION_TOKEN":
+                            sessionToken = (string)pair.Value;
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", nameof(properties));
+                    }
+                }
+            }
+
+            return sessionToken;
+        }
+
+        private static SecureString ToSecureString(string str)
```

Comment:
I'm not sure that we need to have this logic here. We already have the same method in UsernamePasswordCredential. I think we can create a helper for that target or move this method to BsonUtils. Thoughts? 
I will be ok with any solution here.

Reply:
As you can see `UsernamePasswordCredential.ConvertPasswordToSecureString` is private: [link](https://github.com/mongodb/mongo-csharp-driver/blob/719731cb7548ae6c1cfc253c33384084f84fb3b1/src/MongoDB.Driver.Core/Core/Authentication/UsernamePasswordCredential.cs#L146)
And I wouldn't like to touch existing code, unless there's a crucial necessity. Moreover, my updated `ToSecureString` wouldn't throw on `null` input. So I would like to leave this method, unless you insist.